cmd/uniq/uniq_test.go

```diff
@@ -0,0 +1,315 @@
+		// uniq -empty -num
+		options{
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"2: world\n" +
+			"3,5: \n" +
+			`8: 世
+10: 3
+12: 日本語
+`,
+	},
+	{
+		// uniq -dup
+		options{printDuplicates: true},
+		ttScanLines[0].expected,
+		`hello
+世界
+1
+4
+`,
+	},
+	{
+		// uniq -dup -num
+		options{
+			printDuplicates: true,
+			printLineNumber: true,
+		},
+		ttScanLines[0].expected,
+		`1,4: hello
+6,7: 世界
+9,14: 1
+11,13: 4
+`,
+	},
+	{
+		// uniq -dup -empty
+		options{
+			printDuplicates: true,
+			printEmptyLines: true,
+		},
+		ttScanLines[1].expected,
+		`hello
+
+世界
+1
+4
+`,
+	},
+	{
+		// uniq -dup -empty -num
+		options{
+			printDuplicates: true,
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"1,4: hello\n" +
+			"3,5: \n" +
+			`6,7: 世界
+9,14: 1
+11,13: 4
+`,
+	},
+
+	{
+		// uniq -every
+		options{printEveryOnce: true},
+		ttScanLines[0].expected,
+		`hello
+world
+世界
+世
+1
+3
+4
+日本語
+`,
+	},
+	{
+		// uniq -every -num
+		options{
+			printEveryOnce:  true,
+			printLineNumber: true,
+		},
+		ttScanLines[0].expected,
+		`1,4: hello
+2: world
+6,7: 世界
+8: 世
+9,14: 1
+10: 3
+11,13: 4
+12: 日本語
+`,
+	},
+	{
+		// uniq -every -empty
+		options{
+			printEveryOnce:  true,
+			printEmptyLines: true,
+		},
+		ttScanLines[1].expected,
+		`hello
+world
+
+世界
+世
+1
+3
+4
+日本語
+`,
+	},
+	{
+		// uniq -every -empty -num
+		options{
+			printEveryOnce:  true,
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"1,4: hello\n" +
+			"2: world\n" +
+			"3,5: \n" +
+			`6,7: 世界
+8: 世
+9,14: 1
+10: 3
+11,13: 4
+12: 日本語
+`,
+	},
+}
+
+func cmpLines(a, b []*Line) error {
+	if la, lb := len(a), len(b); la != lb {
+		return fmt.Errorf("%T sizes diverge: %v != %v", a, la, lb)
+	}
+	for i, _ := range a {
+		linea := *a[i]
+		lineb := *b[i]
+		if *linea.text != *lineb.text {
+			return fmt.Errorf("line %d:\n%v !=\n%v",
+				i+1, *linea.text, *lineb.text)
+		}
+
+		if lena, lenb := len(linea.nums), len(lineb.nums); lena != lenb {
+			return fmt.Errorf("line counts diverge: %v != %v", lena, lenb)
+		}
+		for n, _ := range linea.nums {
+			lna := linea.nums[n]
+			lnb := lineb.nums[n]
+			if lna != lnb {
+				return fmt.Errorf("line number index: %d -"+
+					"numbers diverge: %v != %v", n, lna, lnb)
+			}
+		}
+	}
+	return nil
+}
+
+func TestScanLines(t *testing.T) {
+	for i, tScanLines := range ttScanLines {
+		bytesReader := bytes.NewReader(input)
+		scannedLines := scanLines(bytesReader, tScanLines.opts)
+		if scannedLines == nil {
+			t.Fatalf("test index %d: %v %[2]T %[2]v", i, scannedLines)
+		}
+
+		err := cmpLines(scannedLines, tScanLines.expected)
+		if err != nil {
+			t.Fatalf("test index %d: %v", i, err)
+		}
+	}
+}
+
+func getUniqOutput(lines []*Line, opts options) (string, error) {
+	var buf bytes.Buffer
+	var err error
+
+	rEnd, wEnd, err := os.Pipe()
+	if err != nil {
+		return "", err
+	}
+
+	savedStdout := os.Stdout
```

Comment:
that's not idiomatic Go. The correct way would be simple define `uniq` something like:

```go
func uniq(stdout io.Writer, linesp []Line, opts options) error { /* code here */ }
```

And the use like this in main:
```go
linesp, err := scanLines(os.Stdin, opts)
if err != nil {
        fmt.Fprintf(os.Stderr, err.Error())
        os.Exit(1)
}
 err = uniq(os.Stdout, linesp, opts)
if err != nil {
        fmt.Fprintf(os.Stderr, err.Error())
        os.Exit(1)
} 
```

This way in your tests you can mock the stdout with:
```go
var buf bytes.Buffer
err := unit(&buf, linesp, opts)
if err != nil {
        t.Fatal(err)
}
out := string(buf.Bytes())
```